### PR TITLE
Add support for 32KB multi_tensor_apply kernel arguments

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134373

## Benchmark

On H100 SXM (HBM2e, 500W TDP), CUDA Toolkit=12.2, Driver Version=535.154.05, with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa) (`torch._foreach_copy_`):

**Baseline**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp0g_x4sys
device ms: 0.891, cpu ms: 7.200
memory bandwidth: 1457.727 GB/s
```

Single iteration trace:
image

**This PR**
```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_tmp3jqiugli
device ms: 0.683, cpu ms: 6.745
memory bandwidth: 1902.010 GB/s
```

Single iteration trace:
image

## Binary Size and Kernel Specialization
The binary size for `libtorch_cuda.so` increased 6MB (243MB -> 249MB).

```
// NOTE: [32KB kernel argument size support]
// 32KB kernel argument size support has three requirements:
// - CUDART_VERSION >= 12010
// - Driver version >= 530
// - GPU arch >= VOLTA
//
// Due to minor version compatibility, it possible for binaries built with
// CUDART_VERSION >= 12010 to run with driver version < 530. Since driver
// version can only be checked at runtime, if CUDART_VERSION >= 12010, we have
// to build both 4KB and 32KB kernels and determine the appropriate kernel to
// dispatch at runtime.
//
// - If CUDART_VERSION < 12010, only 4KB kernels will be instantiated.
//
// - If CUDART_VERSION >= 12010:
//   - Host code:
//     - We always instantiate the launching stub for both 4KB and 32KB kernels.
//   - Device code:
//     - If __CUDA_ARCH__ >= 700, we always instantiate both 4KB and 32KB
//     kernels.
//     - If __CUDA_ARCH__ < 700, it's not possible to even compile an empty
//     32KB kernel (formal parameter space overflowed). Thus, we only
//     instantiate a declaration for 32KB kernels. This is valid as long as the
//     declaration-only kernel is not launched.
//
// - At runtime, we dispatch to the 32KB kernel if driver version >= 530 and
// GPU arch >= VOLTA.
//
// - TODO(yifu): once there's a CUDART version that is not compatible with any
// driver version below 530, we can determine at compile time to not compile
// the kernels for 4KB kernel argument size.
//
// https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/
```

### CODE REVIEW

**Review by eqy — APPROVED** [review]

**Comment by eqy** [comment]
Looks like bazel build is failing for some reason?

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 1]
```diff
@@ -23,6 +23,84 @@ static constexpr int depth_to_max_tensors_scalarlist_of_complex_double[2] = {
...
+//
+// - If CUDART_VERSION >= 12010, kernels for both 4KB and 32KB kernel argument
+// sizes will be built. However, due to CUDA's minor version compatibility,
+// even when CUDART_VERSION >= 12010, the driver may not support 32KB.
+// Therefore, a runtime check is necessary to determine the appropriate kernel
+// to dispatch.
+//
+// - TODO(yifu): once there's a CUDART version that is not compatible with any
+// driver version below 530, we can determine at compile time to not compile
+// the kernels for 4KB kernel argument size.
```
Comment (crcrpar) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:42]: nice comment

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 2]
```diff
@@ -38,60 +116,96 @@ __device__ __forceinline__ void load_store(
...
 
 // note(mkozuki): `n` of 1&2 violate the limit of cuda kernel argument size of
 // 4kb with `c10::complex<double>`
-template <>
-struct TensorListScalarListMetadata<c10::complex<double>, 1> {
-  const void* addresses[1]
-                       [depth_to_max_tensors_scalarlist_of_complex_double[0]];
-  int64_t
-      numel_for_tensor[depth_to_max_tensors_scalarlist_of_complex_double[0]];
+template <bool large_kernel_arg>
```
Comment (crcrpar) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:153]: wouldn't the comment one line above need update?

**Review by crcrpar — APPROVED** [review]
great update, thank you for working on this :)

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 3]
```diff
@@ -14,14 +14,93 @@ static constexpr int64_t kILP = 4;
...
+//
+// https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/
+bool supports_large_kernel_arg() {
+#if !defined(USE_ROCM) && defined(CUDART_VERSION) && CUDART_VERSION >= 12010
+  static std::optional<bool> supports_large_kernel_arg_ = std::nullopt;
+  if (!supports_large_kernel_arg_.has_value()) {
+    int driver_ver = 0;
+    cudaDriverGetVersion(&driver_ver);
+    cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+    *supports_large_kernel_arg_ = (driver_ver >= 12010) && prop->major >= 7;
```
Comment (albanD) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:43]: Driver version 530 has numerical value 12010 in this struct?

**Review by albanD — COMMENTED** [review]
Nice!

FYI @seemethere for the non-trivial binary size increase. I think it is worth it given the speedup and we can in theory recover it in the future when nvidia stops supporting driver 530.

**Comment by eqy** [comment]
Out of curiosity, what was the fix for the bazel build issue, was it actually build related or because that particular job was running on an unsupported GPU?

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 4]
```diff
@@ -14,14 +14,93 @@ static constexpr int64_t kILP = 4;
...
+//
+// - TODO(yifu): once there's a CUDART version that is not compatible with any
+// driver version below 530, we can determine at compile time to not compile
+// the kernels for 4KB kernel argument size.
+//
+// https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/
+bool supports_large_kernel_arg() {
+#if !defined(USE_ROCM) && defined(CUDART_VERSION) && CUDART_VERSION >= 12010
+  static std::optional<bool> supports_large_kernel_arg_ = std::nullopt;
+  if (!supports_large_kernel_arg_.has_value()) {
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:39]: looks like a noob q: when would line 39 ever be False if line 38 sets it to null? Is `supports_large_kernel_arg_1 accessible from outside this function?

**Comment by yifuwang** [comment]
> was it actually build related or because that particular job was running on an unsupported GPU

@eqy it's both. The bazel build was using an M40 GPU and the feature actually required [Volta and above](https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/). We have to handle __CUDA_ARCH__ < 700 specially at compile time, which wasn't done in the previous version. I'll include a comment to describe this.

**Review by janeyx99 — COMMENTED** [review]
Thanks for pursuing this change!

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh (hunk 4 above)
Comment (janeyx99) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:39]: OOOOH i missed the static. Very cool. It has the same effect as function caching then.

**Review by janeyx99 — APPROVED** [review]

**Comment by janeyx99** [comment]
@yifuwang I think this PR is good to go

**Comment by yifuwang** [comment]
The previous merge attempt failed due to an MSVC build issue. Trying to merge again with a possible fix.
```
// MSVC has a problem with constexpr and can't handle passing them to templates
// as arguments. We need to replace it with const static.
// https://github.com/state-spaces/mamba/issues/12#issuecomment-1848835662
```

@pytorchbot merge

**Comment by malfet** [comment]
I think one needs to add driver version check before using the feature..

**Comment by yifuwang** [comment]
@mafet we already do a driver version check in this PR - when building with eligible toolkit for eligbile CUDA_ARCH, we build the 32KB kernels in addition to 4KB kernels. At runtime, we only choose the 32KB kernels if the driver version is eligible. Otherwise, we fallback to the 4KB kernels.

It's having a very weird issue when running with minor version compatibility (e.g. 525 + 12.3). I actually verifed the PR with this combination of driver + toolkit version and did not run into issues. However, digging into https://github.com/pytorch/pytorch/issues/135126, I found that if we call `cudaGetDeviceCount` before we perform a device allocation, the device allocation would fail with "operation not supported" error. If we perform a device allocation first, the failure goes away. **The failure didn't even involve `multi_tensor_apply`, so I suspect that the driver is unhappy about what's in nvfatbin and it only happen in certain scenarios.**

This is unfortunate. If this PR's approach doesn't work, it means that we can only enable this feature when the user tells us the driver version the binary would run on at compile time. It also means that we can not enable this feature in our pre-built binaries (unless we build another version for 530+) until the next major toolkit version.

**Pull request closed without merging**